Unit conversion plugin

jeremypw opened the pull request:
Fixes #114 

Allows conversion between measurement units (*not* variable conversions like currency)

 - Uses syntax `<number><unit> => <unit>`  or `<unit> => <unit>`  (whitespace ignored)
 - If  `<number>` is omitted it is assumed to be "1.0"
 - Allows unit synonyms to be specified (e.g. "click" for kilometer)
 - Allows multiple abbreviations for each unit.
 - Allows units to have a translatable description (which can be matched on)
 - Allows each unit to be preceded by a (metric) prefix
 - Allows each unit to be followed by a dimension or 2 or 3 (may not always make sense in which no match)
 - Handles ambiguous units like "gallon" by offering multiple results

Could maybe copy the database of units from e.g. the `units` package and parse this.  This would give access to a very wide range of units (many of which are obscure) but the philosophy of that database is different - each unit is defined in terms of only one other unit (except some units are fundamental or are multidimensional) and prefixes are handled separately so determining the  applicable factor is not at all straightforward.

At present uses a hard-coded array of Unit structs that has a similar, but simpler structure to `definitions.units` from the `Units` package - each unit is defined in terms of only one other unit using simple integers or fractions where possible within a unit system.  Each system/type of unit has a base_unit and the base_units of non-metric systems are all linked to the corresponding base metric unit.  Links between the required units are traced to calculate the necessary conversion factor.   This makes the code more complex but the table shorter.  The calculation is no longer passed out to `bc`.

At present links are traced by simply iterating through the units as the table is not very large.  In the unlikely event of performance becoming an issue the table could be used to construct a Map when a conversion is first requested.

The number of entries in the table is reduced by handling (metric) prefixes and dimensions (square, cubic) separately.

Unit testing has been included that includes validating the data format as well as testing the results of representative user input.

The total number of conversions is (theoretically) (`<no of prefixes> * <no of entries> * <no of dimensions) ^ 2` although a *lot* of these will be invalid (e.g. meter => kg) or not very useful (e.g. megainches).  This is currently (20 *  34 * 3) ^ 2 = approx 4 million.

Not handled so far includes (to be decided which, if any need be addressed):
 - Conversions involving offset conversion e.g. degC => degF
 - Conversions involving non-linear conversion e.g. Gas Mark => degC
 - Conversions involving reciprocals e.g. ohm => mho
 - Conversion with variable factors e.g. Between different currencies.

jeremypw commented:
@danrabbit I am not sure that the current approach is sufficiently scalable so any suggestions as to which external conversion database or package we could leverage would be welcome.  Not sure whether we want to cover all possible conversions or just stick to common units?  An advantage of the current method (maybe the only one) is that it handles ambiguous units.

jeremypw commented:
![Screenshot from 2022-07-17 19 33 44](https://user-images.githubusercontent.com/10513844/179419964-6489b229-c6d2-4f30-9fc7-1f52043d7da0.png)

jeremypw commented:
Could probably reduce the number of factors by including dimensional analysis - i.e. since 1 liter = 1dm3, since 1dm = 0.1m we already know 1 liter = (0.1)^3 m3.  Its a balance between a long table and more complex code.

jeremypw commented:
Opening for review in order get ideas about desired scope and acceptable formats of input etc.

peteruithoven commented on src/synapse-plugins/converter-plugin/converter-data.vala:
```diff
@@ -0,0 +1,214 @@
+        {"nano", "n", 1E-9},
+        {"pico", "p", 1E-12},
+        {"femto", "f", 1E-15},
+        {"atto", "a", 1E-24},
```
```suggestion
        {"atto", "a", 1E-18},
```
See: https://www.wolframalpha.com/input?i=1atto

peteruithoven commented on src/synapse-plugins/converter-plugin/converter-data.vala:
```diff
@@ -0,0 +1,214 @@
+        {"tera", "T", 1E12},
+        {"giga", "G", 1E9},
+        {"mega", "M", 1E6},
+        {"kilo", "k", 1000},
```
```suggestion
        {"kilo", "kg", 1000},
```

See: https://www.wolframalpha.com/input?i=1+kilo

jeremypw replied:
I do not think this is right - "kg" is a specific unit of mass whereas "k" is a prefix for "1000" it is also used for "km" (kilo-meter) for example.  I know "kilo" is also commonly used as an abbreviation for "kilogram" but that is not its function here. It might be an idea to add another entry to UNITS to accomodate this abbreviation.

jeremypw replied:
I have now added "kilogram" as an entry in UNITS with abbreviation "kilo".

peteruithoven replied:
> I do not think this is right - "kg" is a specific unit of mass whereas "k" is a prefix for "1000" it is also used for "km" (kilo-meter) for example

You are absolutely right, I got confused.

peteruithoven commented on src/synapse-plugins/converter-plugin/converter-data.vala:
```diff
@@ -0,0 +1,214 @@
+        {UnitType.TIME, UnitSystem.SI, "day", "da|d", NC_(TIME, "day"), "24", "hour"},
+        {UnitType.TIME, UnitSystem.SI, "week", "wk", NC_(TIME, "week"), "7", "day"},
+        {UnitType.TIME, UnitSystem.SI, "fortnight", "", NC_(TIME, "fortnight"), "14", "day"},
+        {UnitType.TIME, UnitSystem.SI, "commonyear", "yr|y|calendaryear", NC_(TIME, "Common year"), "365", "day"},
```
```suggestion
        {UnitType.TIME, UnitSystem.SI, "commonyear", "calendaryear|year|yr|y", NC_(TIME, "Common year"), "365", "day"},
```
Shouldn't this have  `year` as an abbreviation? 
I would also reorder the abbreviations from long to small like you've done elsewhere.

peteruithoven commented on src/synapse-plugins/converter-plugin/converter-data.vala:
```diff
@@ -0,0 +1,214 @@
+        // Velocity units  - at present treated as separate unit system although could be calculated from components
+        {UnitType.VELOCITY, UnitSystem.SI, "meterpersecond", "m/s", NC_(TIME, "meters per second"), "1", ""}, // Fundamental
+        {UnitType.VELOCITY, UnitSystem.SI, "lightspeed", "c", NC_(TIME, "speed of light"), "299792458", "meterpersecond"},
+        {UnitType.VELOCITY, UnitSystem.SI, "kilometersperhour", "kph", NC_(TIME, "kilometers per hour"), "1000/3600", "meterpersecond"},
```
```suggestion
        {UnitType.VELOCITY, UnitSystem.SI, "kilometersperhour", "km/h|kph", NC_(TIME, "kilometers per hour"), "1000/3600", "meterpersecond"},
```

peteruithoven commented on src/synapse-plugins/converter-plugin/converter-data.vala:
```diff
@@ -0,0 +1,214 @@
+        {UnitType.VELOCITY, UnitSystem.SI, "meterpersecond", "m/s", NC_(TIME, "meters per second"), "1", ""}, // Fundamental
+        {UnitType.VELOCITY, UnitSystem.SI, "lightspeed", "c", NC_(TIME, "speed of light"), "299792458", "meterpersecond"},
+        {UnitType.VELOCITY, UnitSystem.SI, "kilometersperhour", "kph", NC_(TIME, "kilometers per hour"), "1000/3600", "meterpersecond"},
+        {UnitType.VELOCITY, UnitSystem.SI, "milesperhour", "mph", NC_(TIME, "miles per hour"), "1609.34/3600", "meterpersecond"},
```
```suggestion
        {UnitType.VELOCITY, UnitSystem.SI, "milesperhour", "m/h|mph", NC_(TIME, "miles per hour"), "1609.34/3600", "meterpersecond"},
```

jeremypw replied:
I've not seen "m/h" used as an abbreviation but it does no harm to add it I guess (it is unlikely to be meters per hour)

peteruithoven commented on src/synapse-plugins/converter-plugin/converter-data.vala:
```diff
@@ -0,0 +1,214 @@
+        {UnitType.TIME, UnitSystem.SI, "gregorianyear", "yr|y", NC_(TIME, "Gregorian year"), "366.2425", "day"},
+        {UnitType.TIME, UnitSystem.SI, "islamicyear", "yr|y", NC_(TIME, "Islamic year"), "354", "day"},
+        {UnitType.TIME, UnitSystem.SI, "islamicleapyear", "yr|y", NC_(TIME, "Islamic leap year"), "354", "day"},
+        {UnitType.TIME, UnitSystem.SI, "decade", "", NC_(TIME, "decade"), "10", "year"},
+        {UnitType.TIME, UnitSystem.SI, "century", "", NC_(TIME, "century"), "100", "year"},
+        {UnitType.TIME, UnitSystem.SI, "millenium", "", NC_(TIME, "millenium"), "1000", "commonyear"},
```
```suggestion
        {UnitType.TIME, UnitSystem.SI, "decade", "", NC_(TIME, "decade"), "10", "commonyear"},
        {UnitType.TIME, UnitSystem.SI, "century", "", NC_(TIME, "century"), "100", "commonyear"},
        {UnitType.TIME, UnitSystem.SI, "millenium", "", NC_(TIME, "millenium"), "1000", "commonyear"},
```
I'm not sure how this works yet, but it should probably all be `commonyear`, as that's the main name?

jeremypw replied:
Quite right - the link should be a uid - I didn't update it when I changed the uid to commonyear.

peteruithoven reviewed (commented):
First few things I noticed. Still need to look at `converter-plugin-backend.vala` and `converter-plugin.vala`. Might take me a while.

peteruithoven commented:
I am curious btw, why wouldn't we use an existing library like `qalc` for this? No control over the input syntax and in some way harder to maintain because of a dependency on external code?

peteruithoven commented on src/synapse-plugins/converter-plugin/converter-plugin-backend.vala:
```diff
@@ -0,0 +1,360 @@
+
+        public ResultData[] get_conversion_data (string query_string) {
+            var input = query_string.replace (" ", "").replace (",", ".").replace ("|", "");
+            var matched = convert_regex.match (input);
```
The `convert_regex` is quite complex. Would it make sense for performance to add a simple contains "=>" check?

jeremypw replied:
OK, I can do that.

jeremypw replied:
Done - although if it is decided to allow more possible formats or alternatives to "=>" this will need to be reverted.

peteruithoven commented on src/synapse-plugins/converter-plugin/converter-plugin-backend.vala:
```diff
@@ -0,0 +1,360 @@
+             */
+            try {
+                convert_regex = new Regex (
+                    """^\d*.?\d*[a-zA-Z\/ ]+(2|3)?=>[a-zA-Z\/ ]+(2|3)?$""",
```
Why `d*.?\d` and not simply `\d*`

jeremypw replied:
My understanding was that`\d` matched a "digit" not a decimal point? So I included a match for a single decimal point.  I'll check if it can be omitted. Or I guess I could use `[0-9.]`?  My knowledge of regex is a bit rudimentary.

jeremypw replied:
I tried it with just `\d*` and decimal points were not recognised.

jeremypw replied:
Done

peteruithoven replied:
> So I included a match for a single decimal point.

Ahh. But a `.` in this context means any character. You'd need to escape it (`\.`) to just match with `.`'s. 
Just the `\d` should not match `.`'s, see: https://regex101.com/r/sy5EHo/1
I would use `[\d.]`. Note that inside `[]`'s the dot only matches with dots and no escape is needed.

Example with `[\d.]`: https://regex101.com/r/jfp9SF/1

jeremypw replied:
That's what I originally thought re `\d` but now all the test units are passing - even the ones including a decimal point even though I've taken out the match for the decimal point so I am not sure what is going on :shrug:  Anyway I'll put the decimal point regex back in according to your suggestion.

peteruithoven commented on src/synapse-plugins/converter-plugin/converter-plugin-backend.vala:
```diff
@@ -0,0 +1,360 @@
+             */
+            try {
+                convert_regex = new Regex (
+                    """^\d*.?\d*[a-zA-Z\/ ]+(2|3)?=>[a-zA-Z\/ ]+(2|3)?$""",
```
The `[a-zA-Z\/ ]`'s seems to support including spaces, even though you're removing spaces beforehand?

jeremypw replied:
I was wanting to support input units like "US cup" so I guess I should not be removing spaces before matching.

jeremypw replied:
Fixed

peteruithoven commented on src/synapse-plugins/converter-plugin/converter-plugin-backend.vala:
```diff
@@ -0,0 +1,360 @@
+             */
+            try {
+                convert_regex = new Regex (
+                    """^\d*.?\d*[a-zA-Z\/ ]+(2|3)?=>[a-zA-Z\/ ]+(2|3)?$""",
```
`(2|3)` could simply be `[23]`? The `|` syntax seems more appropriate when parts contain multiple characters?

jeremypw replied:
Agreed.

jeremypw replied:
Done

jeremypw commented:
Re `qalc` both reasons you suggested are right. The user input is not known to be a calculation or conversion so it is helpful to have control on how it is parsed and how ambiguities are dealt with.  This is only intended to be `quick` limited tool for common conversions (I included some more obscure units for demonstration/development purposes), in the same way that the calculator plugin is not intended to emulate the calculator app. I also wanted to avoid another dependency.   A lot of the code is involved in parsing the input and you would probably have to keep a fair amount of that to get the required input for an external library (and you would have to parse the output).

There may be some scope for slimming down the code (especially if you reduce the flexibility regarding input or scope of conversions) but I think the size is manageable.  Refinements like handling complex conversions by dimensional analysis would increase the code considerably and are probably outside the scope of this plugin.

The user can always type into Google or install e.g. Qalculate!  for advanced stuff.

jeremypw commented:
@peteruithoven Thanks for your efforts in reviewing this.  I have addressed your comments so far and look forward to your further comments.

jeremypw commented:
Matching with decimal point and unit description with embedded space:

![Screenshot from 2022-07-25 16 02 04](https://user-images.githubusercontent.com/10513844/180809895-760f4bd0-24bd-4c1b-8275-6ba6f3040cf8.png)

peteruithoven commented on src/synapse-plugins/converter-plugin/converter-plugin-backend.vala:
```diff
@@ -0,0 +1,372 @@
+
+            parent_dimension = 1;
+            var ultimate_parent1 = parent.uid;
+            parent = u2;
+            debug ("finding root of %s - start dimension %i, start factor2 %f", u2.uid, dim2, factor2);
+            while (parent != null &&
+                   parent.base_unit != "" &&
+                   (!same_system || parent.system == u2.system)) {
+
+                parent = find_parent_unit (parent.base_unit, out parent_dimension);
+                var pfactor = parent.get_factor ();
+                debug ("parent2 %s parent_dimension2 %i, parent_factor2 %f",
+                    parent.uid, parent_dimension, pfactor
+                );
+                dim2 *= parent_dimension;
+                debug ("Dim2 now %i", dim2);
+                for (int i = 0; i < dim2; i++) {
+                    factor2 *= pfactor;
+                    debug ("Factor2 now %f", factor2);
+                }
+            }
```
This is very similar to the parent1 logic, maybe it would be better to move it to a function which you could reuse twice?

peteruithoven commented on src/test/ConverterPluginTest.vala:
```diff
@@ -0,0 +1,110 @@
+        run_conversion (input, result, true);
+    }
+
+    static void assert_throw (string input, double result) {
+        run_conversion (input, 0, false);
+    }
```
The result parameter isn't used? Since we expect a throw I'm not sure it makes sense to specify a result?
(If we remove it we might want to separate it a bit more in `main`)

jeremypw replied:
Removed.

peteruithoven reviewed (changes requested):
I finally found some time to go through it. It's a impressive bit of logic.
I only found some minor things. I'm curious how you feel about using the regular expression for parsing. While it would complicate the regular expression it would reduce the amount of parsing logic by quite a lot. 

Two future improvement ideas:
- Supporting plural versions (e.g. `2 meters => mm`). Could be as simple as adding the plural version to the `abbreviations`, or maybe ignoring an `s` somehow. 
- Supporting `->` as well?

jeremypw commented:
@peteruithoven Thanks for the review - I'll go through your suggestions later today. Regarding plurals, this would be difficult to accomodate in a localizable way as there is no framework for reverse translation.  Ignoring a trailing 's' would fail for e.g. 'inch'.  So I would prefer to leave that for a future refinement if a reasonable solution can be found.  Alternatives to `=>` can be accomodated as long as they are unlikely to appear in other types of query.

peteruithoven commented:
Ah yeah the future improvement ideas should not be a part of this PR, let's get this out there first :) 
I'll probably create issues for them when this gets merged. 
Good points on the plurals.

jeremypw commented on src/synapse-plugins/converter-plugin/converter-data.vala:
```diff
@@ -0,0 +1,203 @@
+
+    struct Unit {
+        public UnitType type; // Units of different types cannot be interconverted
+        public UnitSystem system; // e.g. SI, Imperial, Chinese
```
Yet ...  I was looking at incorporating Chinese units but perhaps that can be left for now.

jeremypw commented on src/synapse-plugins/converter-plugin/converter-data.vala:
```diff
@@ -0,0 +1,203 @@
+    struct Unit {
+        public UnitType type; // Units of different types cannot be interconverted
+        public UnitSystem system; // e.g. SI, Imperial, Chinese
+        public string uid; // Unique identifier for the purposes of the converter - not official
```
Just that it does not have to correspond to anything used in the real world (although it will be recognized if entered).  The important thing is that it is unique.

jeremypw commented on src/synapse-plugins/converter-plugin/converter-data.vala:
```diff
@@ -0,0 +1,203 @@
+        {UnitType.DIMENSION, UnitSystem.IMPERIAL, "cmile", "mi|cmi|mile", NC_(LENGTH, "country mile"), "2200", "yard"},
+
+        // Area units
+        {UnitType.DIMENSION, UnitSystem.IMPERIAL, "iacre", "acre", NC_(LENGTH, "International acre"), "10", "chain2"},
+        {UnitType.DIMENSION, UnitSystem.IMPERIAL_US, "usacre", "acre", NC_(LENGTH, "US acre"), "10", "surveyorschain2"},
```
Yes, good spot.

jeremypw commented on src/synapse-plugins/converter-plugin/converter-data.vala:
```diff
@@ -0,0 +1,203 @@
+        {UnitType.TIME, UnitSystem.SI, "millenium", "", NC_(TIME, "millenium"), "1000", "commonyear"},
+
+        // Velocity units  - at present treated as separate unit system although could be calculated from components
+        {UnitType.VELOCITY, UnitSystem.SI, "meterpersecond", "m/s", NC_(TIME, "meters per second"), "1", ""}, // Fundamental
+        {UnitType.VELOCITY, UnitSystem.SI, "lightspeed", "c", NC_(TIME, "speed of light"), "299792458", "meterpersecond"},
+        {UnitType.VELOCITY, UnitSystem.SI, "kilometersperhour", "km/h|kph", NC_(TIME, "kilometers per hour"), "1000/3600", "meterpersecond"},
+        {UnitType.VELOCITY, UnitSystem.SI, "milesperhour", "m/h|mph", NC_(TIME, "miles per hour"), "1609.34/3600", "meterpersecond"},
+        {UnitType.VELOCITY, UnitSystem.SI, "mach", "", NC_(TIME, "Mach (speed of sound)"), "331.46", "meterpersecond"},
```
Yes, good spot.

jeremypw commented on src/synapse-plugins/converter-plugin/converter-plugin-backend.vala:
```diff
@@ -0,0 +1,372 @@
+
+            try {
+                convert_regex = new Regex (
+                    """\d*[a-zA-Z\/ ]+[23]?=>[a-zA-Z\/ ]+[23]?""",
```
The tests still pass with your regex so I have used it. I'll look into parsing with regex later.  (I am not an expert with regex).

jeremypw commented on src/synapse-plugins/converter-plugin/converter-plugin-backend.vala:
```diff
@@ -0,0 +1,372 @@
+                return results;
+            }
+
+            var input = query_string.replace (",", ".").replace ("|", "").strip ();
```
Its probably unnecessary as an entry with that character probably isn't a valid conversion request anyway.  It was probably because that character is used to separate alternative abbreviations. I'll take it out if the tests still pass.

jeremypw commented on src/synapse-plugins/converter-plugin/converter-plugin-backend.vala:
```diff
@@ -0,0 +1,372 @@
+                // Parse input into a number and two unit match arrays
+                // Some abbreviations are ambiguous (used in >1 system) so get all possible matching units
+
+                var parts = input.split ("=>", 2);
+                var num_s = parts[0];
+                // Isolate leading number (First \0 truncates string)
+                num_s.canon ("1234567890.", '\0');
+                if (num_s.contains ("..")) { // Multiple decimal points are incorrectly parsed to double
+                    return results;
+                }
+
+                if (num_s.length > 0) { // If leading number omitted, assume it to be 1.0
+                    num = double.parse (num_s);
+                }
+
+                // Get user input units; cleanup excess whitespace. Note: more than two consecutive embedded spaces
+                // will prevent a match.
+                string unit1_s = (parts[0].slice (num_s.length, parts[0].length)).strip ().replace ("  ", " ");
+                string unit2_s = parts[1].strip ().replace ("  ", " ");
+
+                // Split each unit into prefix, base and dimension
+                get_prefix_and_dimension (unit1_s, out prefix1_s, out prefix1, out dimension1);
+                get_prefix_and_dimension (unit2_s, out prefix2_s, out prefix2, out dimension2);
```
That conversion works (with your regex).  I'll look into regex parsing later.

jeremypw commented on src/synapse-plugins/converter-plugin/converter-plugin-backend.vala:
```diff
@@ -0,0 +1,372 @@
+            var match = unit_s;
+            // Test match whole unit
+            foreach (string id in ids) {
+                if (match == id) {
```
Need to be careful about prefixes such as "m (milli)" and "M (mega)" but you are right - those examples do not work.  Need to fix and add to tests

jeremypw replied:
I have now altered the match to make it case insensitive.

jeremypw commented on src/synapse-plugins/converter-plugin/converter-plugin-backend.vala:
```diff
@@ -0,0 +1,372 @@
+                }
+            }
+
+            if (prefix != "") {
+                //Test match without prefix
+                match = unit_s[prefix.length : unit_s.length];
+                foreach (string id in ids) {
+                    if (match == id) {
+                        debug ("unit less prefix matches");
+                        use_prefix = true;
+                        return true;
+                    }
+                }
+            }
+
+            if (dimension > 1) {
+                //Test match without dimension
+                match = unit_s[0 : -1];
+                foreach (string id in ids) {
+                    if (match == id) {
+                        debug ("unit less dimension matches");
+                        use_dimension = true;
+                        return true;
+                    }
+                }
+            }
+
+            if (prefix != "" && dimension > 1) {
+                //Test match without either prefix or dimension
+                match = unit_s[prefix.length : -1];
+                foreach (string id in ids) {
+                    if (match == id) {
+                        debug ("unit less both matches");
+                        use_prefix = use_dimension = true;
+                        return true;
+                    }
+                }
+            }
```
At this point we still do not know what `prefix` and `dimension` are so we have to test.

peteruithoven replied:
I might misunderstand the code but didn't `get_prefix_and_dimension` already figure out what the `prefix` and `dimension` are?

What if you replace `prefix` and `dimension` with `""` in `unit_s` and only then try to match it to the `ids`? 

Again I probably misunderstand but why are the `use_prefix` and `use_dimension` needed? Can't we just use the prefix/dimension when they are not empty?

jeremypw replied:
OK, I'll rethink this tomorrow.

jeremypw replied:
OK, so the logic is a little subtle here.  I think we need `use_prefix` because some units e.g. "minute" start with an SI prefix ("m") but it is not actually a prefix and is not used.  We need to distinguish this from cases like "mm" where the prefix "m" *is* used as part of the factor.  I think you may right however that "use_dimension" can be discarded since no units have a uid, abbreviation or description ending in a number (at the moment) so if we find one then it must be used.

jeremypw replied:
Just checked through the `units` database and there are one or two obscure units like "mu0" ending in a number but I think the scope of this tool is likely to be restricted to commonly used units and I cannot think of any that end in a number.

jeremypw replied:
I have now simplified the check considerably - thanks.

peteruithoven replied:
Ah I think I understand the problem. The simplification looks great

jeremypw commented:
@peteruithoven I think I have addressed most of your points, with the exception of regex parsing. It might take me some time to work out how to do that so you might consider merging this or pushing a commit if you have time.

peteruithoven reviewed (approved):
Looks great. Let's, like you suggested, lets tackle using regex for parsing separately.
